### ISSUE TITLE
Toolchains feature does not appear to treat Java as backwards compatible when selecting JDK by vendor

### ISSUE THREAD

**Issue by mikehearn** [gradle/gradle#17444]
Gradle does not seem to do version matching correctly when a vendor match is required.

### Expected Behavior

I have compiled OpenJDK 17 myself with a custom `java.vendor` string, which I have verified by running `java -XshowSettings`. I then use this code to try and select it:

```kotlin
tasks.withType<org.jetbrains.kotlin.gradle.dsl.KotlinJvmCompile>().configureEach {
    println(javaToolchains.compilerFor {
        // languageVersion.set(JavaLanguageVersion.of(17))
        vendor.set(JvmVendorSpec.matching("hydraulic"))
    }.get())
    kotlinOptions.jdkHome = "/Users/mike/Hydraulic/hdk/build/hdk/Contents/Home"
}
```

The `gradle.properties` file contains this:

`org.gradle.java.installations.paths=/Users/mike/path/to/my/jdk`

and `gradle toolChains` shows it in the list with the correct vendor name. So far, so good, although it would be nice if I could just specify the path to the JDK I want instead of needing all this indirection.

However, there are two problems:

1. If no version is specified at all I get this: `Cannot query the value of this provider because it has no value available.`. The docs don't make it clear (in my view) that a version number *must* be specified, and it's not a very intuitive requirement. I control the JDKs that Gradle can see, so I just want it to pick the highest version with the right vendor, but it doesn't do that.
2. If I specify version 16 and my JDK is actually version 17, it still doesn't work. I get this: `No compatible toolchains found for request filter`

The second problem is truly puzzling because Java 17 can run Java 16 code, so if I request 16 and all I have is 17, that should match. I don't see why I should be forced to hard-code the exact version of Java I want simply to use a specific line of JDKs from a specific vendor. Again, the fact that 17 is not considered a suitable match for 16 doesn't seem to be explained in the docs, nor in the error messages.

Possible solutions:

1. Change the matching logic so Vendor V, Version N matches a request for Vendor V, Version N-1
2. Allow versionless JDK requests and then pick the highest.
3. Update the docs to call out this limitation.

The ideal would be 1+2.

**Comment by ilya-g** [comment]
> Change the matching logic so Vendor V, Version N matches a request for Vendor V, Version N-1

I don't think it should be the default behavior when you're requesting a specific version.
Suppose you want to ensure that your app/library can run on Java 16. For that, you query a toolchain for JDK 16 and use its launcher for running tests. It would be unfortunate if Gradle selected JDK 17 instead and gave false confidence that the tests can run on JDK 16.

I agree that we should have more flexibility in querying toolchains: to be able to query them by a more complex predicate than an exact version match. I described my use cases for that in #18894.

**Comment by JojOatXGME** [comment]
There is also #16256 about the ability to accept newer JDKs.

**Comment by ov7a** [comment]
Sorry for the late reply.

This is the intended behavior, and there is no plan to change it. Therefore, this issue will be closed.

***

> if I specify version 16 and my JDK is actually version 17, it still doesn't work. I get this: No compatible toolchains found for request filter

It was considered during the initial design of toolchains, but that approach had issues and was rejected.

The error message about the missing version was updated already.
See also:
* https://docs.gradle.org/current/userguide/toolchains.html#sec:configuring_toolchain_specifications
and 
* https://github.com/gradle/gradle/issues/16256
